torch/onnx/symbolic.py

```diff
@@ -1090,3 +884,133 @@ def symbolic_flattened_wrapper(g, input, *args):
         symbolic_flattened_wrapper
     )
     return tuple(o for o in outputs)
+
+
+def RNN_variant_symbolic_builder(
+        variant, input_size, hidden_size, num_layers, batch_first, dropout, bidirectional, **kwargs):
+    def symbolic(g, input, all_weights, initial_states, batch_sizes):
+        if batch_first:
+            return _unimplemented("RNN/GRU/LSTM", "batch_first")
+        if dropout and kwargs['train']:
+            return _unimplemented("RNN/GRU/LSTM", "dropout in training mode")
+
+        unidirectional = not bidirectional
+
+        prev_output = input
+
+        if variant == 'RNN' or variant == 'GRU':
+            h0 = initial_states
+            h_outs = []
+        elif variant == 'LSTM':
+            h0, c0 = initial_states
+            h_outs = []
+            c_outs = []
+
+        sequence_lens = unused(g) if batch_sizes is None else batch_sizes
+
+        for i in range(num_layers):
+            if unidirectional:
+                if variant == 'RNN':
+                    weight_ih, weight_hh, bias_ih, bias_hh = all_weights[i]
+                elif variant == 'GRU':
+                    # pytorch is reset, input, hidden
+                    # onnx is    input, reset, hidden
+                    weight_ih, weight_hh, bias_ih, bias_hh = \
+                        [reform_weights(g, w, hidden_size, [(1, 2), (0, 1), (2, 3)]) for w in all_weights[i]]
+                elif variant == 'LSTM':
+                    # pytorch is input, forget, cell, output.
+                    # onnx is    input, output, forget, cell.
+                    weight_ih, weight_hh, bias_ih, bias_hh = \
+                        [reform_weights(g, w, hidden_size, [(0, 1), (3, 4), (1, 3)]) for w in all_weights[i]]
+                bias_concat = g.op('Concat', bias_ih, bias_hh, axis_i=0)
+
+                # Unidirectional weights in PyTorch omit the
+                # directionality dimension (dim=0), whereas in ONNX it
+                # is required even for unidirectional RNN; thus,
+                # insert a dimension of size at dim=0.
+                weight_ih = g.op('Unsqueeze', weight_ih, axes_i=[0])
+                weight_hh = g.op('Unsqueeze', weight_hh, axes_i=[0])
+                bias_concat = g.op('Unsqueeze', bias_concat, axes_i=[0])
+
+                if variant == 'RNN' or variant == 'GRU':
+                    state_inputs = [
+                        h0 if num_layers == 1 else g.op('Slice', h0, axes_i=[0], starts_i=[i], ends_i=[i + 1])
+                    ]
+                elif variant == 'LSTM':
+                    state_inputs = [
+                        h0 if num_layers == 1 else g.op('Slice', h0, axes_i=[0], starts_i=[i], ends_i=[i + 1]),
```

Comment:
Can you deduplicate this line as well? Something like:

```
state_inputs = [h0 if num_layers == 1 else g.op('Slice', h0, axes_i=[0], starts_i=[i], ends_i=[i + 1])]
if variant == 'LSTM':
  state_inputs.append(c0 if num_layers == 1 else g.op('Slice', c0, axes_i=[0], starts_i=[i], ends_i=[i + 1]))
```

Reply:
sure, I can apply this type of change in a couple places